### PR TITLE
Feature/dsl/user

### PR DESCRIPTION
**Pull request by GiuliaB97**
Update AutoUserBuilder:

- remove 'isInstanceOf' thanks to the new policy extractors-> update build method
- fix mbf due to the new User interface

Update UserCryptographicInformation

- new algorithm policy

### CODE REVIEW

File: src/main/scala/it/unibo/authsim/library/user/builder/OptionalBuilder.scala [hunk 1]
```diff
@@ -0,0 +1,4 @@
+package it.unibo.authsim.library.user.builder
+
+trait OptionalBuilder[U]:
+  def build():Option[U]
```
Comment (asperan) [inline src/main/scala/it/unibo/authsim/library/user/builder/OptionalBuilder.scala:4]: This trait is redundant with `Builder[T]`, where `T` is `Option[U]`.

File: src/main/scala/it/unibo/authsim/library/user/builder/UserInformationBuilder.scala [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+import it.unibo.authsim.library.dsl.policy.model.Policy
+import it.unibo.authsim.library.dsl.policy.model.StringPolicies.CredentialPolicy
+import it.unibo.authsim.library.user.model.{CryptoInformation, UserInformation}
+import scala.language.postfixOps
+import it.unibo.authsim.library.user.builder.UserBuilder
+
+/**
+ * Class that represent a builder of UserInformation
+ */
+class UserInformationBuilder extends OptionalBuilder[UserInformation]:
```
Comment (asperan) [inline src/main/scala/it/unibo/authsim/library/user/builder/UserInformationBuilder.scala:14]: This class could/should extend `Builder[Option[UserInformation]]`, which also provide a helper method to define building methods:
```suggestion
class UserInformationBuilder extends Builder[Option[UserInformation]]:
```

File: src/test/scala/it/unibo/authsim/library/dsl/attack/builders/BruteForceAttackOptionalBuilderTest.scala [hunk 3]
```diff
@@ -11,7 +11,7 @@ import org.scalatestplus.mockito.MockitoSugar.mock
 
 import scala.concurrent.duration.Duration
 
-class BruteForceAttackBuilderTest extends AnyWordSpec:
+class BruteForceAttackOptionalBuilderTest extends AnyWordSpec:
```
Comment (asperan) [inline src/test/scala/it/unibo/authsim/library/dsl/attack/builders/BruteForceAttackOptionalBuilderTest.scala:14]: Why has this class been renamed?

File: src/test/scala/it/unibo/authsim/library/dsl/attack/builders/DictionaryAttackOptionalBuilderTest.scala [hunk 4]
```diff
@@ -11,7 +11,7 @@ import org.scalatestplus.mockito.MockitoSugar.mock
 
 import scala.concurrent.duration.Duration
 
-class DictionaryAttackBuilderTest extends AnyWordSpec:
+class DictionaryAttackOptionalBuilderTest extends AnyWordSpec:
```
Comment (asperan) [inline src/test/scala/it/unibo/authsim/library/dsl/attack/builders/DictionaryAttackOptionalBuilderTest.scala:14]: Also this class has been renamed, why?

File: src/test/scala/it/unibo/authsim/library/dsl/policy/builders/PolicyOptionalBuilderTests.scala [hunk 5]
```diff
@@ -12,7 +12,7 @@ import org.scalatest.matchers.*
 
 import scala.language.postfixOps
 
-class PolicyBuilderTests extends AnyFlatSpec with should.Matchers:
+class PolicyOptionalBuilderTests extends AnyFlatSpec with should.Matchers:
```
Comment (asperan) [inline src/test/scala/it/unibo/authsim/library/dsl/policy/builders/PolicyOptionalBuilderTests.scala:15]: Also this class has been renamed, why?

File: src/test/scala/it/unibo/authsim/library/dsl/policy/builders/StringPolicyAutoOptionalBuilderTests.scala [hunk 6]
```diff
@@ -10,7 +10,7 @@ import scala.collection.mutable.ListBuffer
 import scala.language.postfixOps
 import scala.util.matching.Regex
 
-class StringPolicyAutoBuilderTests extends AnyFunSuite with BeforeAndAfter:
+class StringPolicyAutoOptionalBuilderTests extends AnyFunSuite with BeforeAndAfter:
```
Comment (asperan) [inline src/test/scala/it/unibo/authsim/library/dsl/policy/builders/StringPolicyAutoOptionalBuilderTests.scala:13]: Also this class has been renamed, why?

**Review by asperan — CHANGES_REQUESTED** [review]

File: src/test/scala/it/unibo/authsim/library/dsl/attack/builders/BruteForceAttackOptionalBuilderTest.scala (hunk 3 above)
Comment (GiuliaB97) [reply src/test/scala/it/unibo/authsim/library/dsl/attack/builders/BruteForceAttackOptionalBuilderTest.scala:14]: Idk : I do not remember having modified your classes

File: src/test/scala/it/unibo/authsim/library/dsl/policy/builders/PolicyOptionalBuilderTests.scala (hunk 5 above)
Comment (GiuliaB97) [reply src/test/scala/it/unibo/authsim/library/dsl/policy/builders/PolicyOptionalBuilderTests.scala:15]: Ok, I definitly have launch a rename and I accidentally to rename all the elements that contain that string. I am sorry.

File: src/test/scala/it/unibo/authsim/library/dsl/attack/builders/DictionaryAttackOptionalBuilderTest.scala (hunk 4 above)
Comment (GiuliaB97) [reply src/test/scala/it/unibo/authsim/library/dsl/attack/builders/DictionaryAttackOptionalBuilderTest.scala:14]: See the previous comment, ah it is probably due to the fixes I made when I introduce the mock class for the Algorithm, or I did a rename that I do not remember and I accidentally rename also your classes.
NP which name should I use?
` DictionaryAttackBuilderTest` is not it? and `BruteForceAttackBuilderTest  `

**Review by asperan — APPROVED** [review]

**Pull request merged**